src/util/strencodings.cpp

```diff
@@ -546,9 +546,18 @@ bool ParseFixedPoint(const std::string &val, int decimals, int64_t *amount_out)
     return true;
 }
 
-void Downcase(std::string& str)
+std::string ToLower(const std::string& str)
 {
-    std::transform(str.begin(), str.end(), str.begin(), [](char c){return ToLower(c);});
+    std::string r;
```

Comment:
Could reserve, same below.

Reply:
What does that mean?

Reply:
See http://www.cplusplus.com/reference/string/string/reserve/.

Reply:
@promag I don't think using `r.reserve(str.size());` necessarily results in faster execution here? I think it depends on the SSO implementation and the length of `str`.

If `str.size() <= 15`  I think **not** using `reserve` is faster when using `libstdc++`. And `str.size() <= 22` for `libc++`.

Am I wrong? :-)

Reply:
Right, but I'm not particularly worried about performance, i think it improves the code.

Reply:
Reserving seems sensible. It would actually slow down the code, @practicalswift ? Crazy. 
I would have never expected that. I'm indifferent on this, though, and opinions are mixed so leaving as is for now.

Reply:
@kallewoof You can test this using the following Google Benchmark snippet:

```
#include <algorithm>
#include <string>

constexpr char ToLower(char c) {
    return (c >= 'A' && c <= 'Z' ? (c - 'A') + 'a' : c);
}

std::string ToLower_1(const std::string& str) {
    std::string r;
    for (auto ch : str) {
        r += ToLower((unsigned char)ch);
    }
    return r;
}

std::string ToLower_2(const std::string& str) {
    std::string r;
    r.reserve(str.size());
    for (auto ch : str) {
        r += ToLower((unsigned char)ch);
    }
    return r;
}

static void BenchToLower_1(benchmark::State& state) {
  for (auto _ : state) {
    const std::string s1{"HELLO"};
    benchmark::DoNotOptimize(s1);
    const std::string s2 = ToLower_1(s1);
    benchmark::DoNotOptimize(s2);
  }
}
BENCHMARK(BenchToLower_1);

static void BenchToLower_2(benchmark::State& state) {
  for (auto _ : state) {
    const std::string s1{"HELLO"};
    benchmark::DoNotOptimize(s1);
    const std::string s2 = ToLower_2(s1);
    benchmark::DoNotOptimize(s2);
  }
}
BENCHMARK(BenchToLower_2);
```

Notice that `ToLower_1` which doesn't `reserve` is slightly faster for strings up to 15 chars for libstdc++ and MSVC and up to 22 chars for libc++.

Reply:
@practicalswift Thanks for that! Very educational. Memory allocations are tricky to optimize, I guess. :)